File: content/snippets/cpp/interview_lcci_17_04_missing_number.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int ans = 0;
        for (int i = 1; i <= (int)nums.size(); i++) {
            ans ^= i;
        }
        for (int x : nums) {
            ans ^= x;
        }
        return ans;
    }
};

int main() {
    Solution sol;
    vector<int> nums1 = {3, 0, 1};
    cout << sol.missingNumber(nums1) << endl;  // 2

    vector<int> nums2 = {9, 6, 4, 2, 3, 5, 7, 0, 1};
    cout << sol.missingNumber(nums2) << endl;  // 8
    return 0;
}

File: content/snippets/python/interview_lcci_17_04_missing_number.py
from typing import List


class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        ans = 0
        for i, x in enumerate(nums, 1):
            ans ^= i ^ x
        return ans


def main():
    sol = Solution()
    print(sol.missingNumber([3, 0, 1]))              # 2
    print(sol.missingNumber([9, 6, 4, 2, 3, 5, 7, 0, 1]))  # 8


if __name__ == "__main__":
    main()
